File: 9- Reverse a linked list/Java/solution3.java
// By dealing with links (recursively):

// Time complexity: O(n)
// Space complexity: O(n)

static class Node{
  int data;
  Node next;

  Node(int data){
    this.data = data;
    this.next = null;
  }

  Node(int data, Node next){
    this.data = data;
    this.next = next;
  }
}

static class LinkedList{
  Node head;

  LinkedList(){
    this.head = null;
  }

  LinkedList(Node head){
    this.head = head;
  }
}

static Node reverseNodes(Node node){
  if(node == null || node.next == null)
    return node;
  Node reversed = reverseNodes(node.next);
  node.next.next = node;
  node.next = null;
  return reversed;
}

static void reverseList(LinkedList list){
  list.head = reverseNodes(list.head);
}

File: 9- Reverse a linked list/C++/solution3.cpp
// By dealing with links (recursively):

// Time complexity: O(n)
// Space complexity: O(n)


struct Node{
  int data;
  Node* next;
};

Node* newNode(int data, Node* next = NULL){
  Node* node = (Node*) malloc(sizeof(Node));
  node->data = data;
  node->next = next;
  return node;
}

struct LinkedList{
  Node* head;
};

LinkedList newLinkedList(Node* head = NULL){
  LinkedList linkedList;
  linkedList.head = head;
  return linkedList;
}

Node* reverseNodes(Node* node){
  if(node == NULL || node->next == NULL)
    return node;
  Node* reversed = reverseNodes(node->next);
  node->next->next = node;
  node->next = NULL;
  return reversed;
}

void reverseList(LinkedList& list){
  list.head = reverseNodes(list.head);
}


